contracts/utils/OptionsPurchaseQueue.sol

```diff
@@ -0,0 +1,660 @@
+                    optionsAmount
+                );
+
+                // Deduct transferred options from allocatedOptions
+                allocatedOptions -= optionsAmount;
+            }
+        }
+
+        // Transfer premiums to the vault
+        if (totalPremiums != 0) asset.safeTransfer(msg.sender, totalPremiums);
+
+        // Clear purchase queue
+        delete purchases[msg.sender];
+        totalOptionsAmount[msg.sender] = 0;
+        // Purchase requests are unblocked
+        vaultAllocatedOptions[msg.sender] = 0;
+
+        emit OptionsSold(msg.sender, totalPremiums, totalOptions);
+
+        return totalPremiums;
+    }
+
+    /************************************************
+     *  OWNER OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice Cancels all purchase requests for a delisted vault
+     * @dev Only callable by the owner. Will revert if options have already been allocated by the vault.
+     * @param vault The vault to cancel all purchases for
+     */
+    function cancelAllPurchases(address vault) external override onlyOwner {
+        // Revert if the vault is still listed
+        require(ceilingPrice[vault] == 0, "Vault listed");
+        // This prevents cancellations after the vault has set its allocation
+        require(vaultAllocatedOptions[vault] == 0, "Vault allocated");
+
+        uint256 totalPremiums;
+        IERC20 asset = IERC20(IRibbonThetaVault(vault).vaultParams().asset);
+        Purchase[] memory purchaseQueue = purchases[vault];
+
+        for (uint256 i; i < purchaseQueue.length; i++) {
+            // Skips request if an instant purchase has already occurred
+            if (purchaseQueue[i].optionsAmount != 0) {
+                // Refund premiums to the buyer
+                asset.safeTransfer(
+                    purchaseQueue[i].buyer,
+                    totalPremiums += purchaseQueue[i].premiums
+                );
+            }
+        }
+
+        emit AllPurchasesCancelled(
+            vault,
+            totalPremiums,
+            totalOptionsAmount[vault]
+        );
+
+        // Clear purchase queue
+        delete purchases[vault];
+        totalOptionsAmount[vault] = 0;
+    }
+
+    /************************************************
+     *  SETTERS
+     ***********************************************/
+
+    /**
+     * @notice Adds a buyer to the purchase queue whitelist
+     * @dev Only callable by the owner
+     * @param buyer The buyer to whitelist
+     */
+    function whitelistBuyer(address buyer) external override onlyOwner {
+        require(buyer != address(0), "!buyer");
+
+        whitelistedBuyer[buyer] = true;
+
+        emit BuyerWhitelisted(buyer);
+    }
+
+    /**
+     * @notice Removes a buyer from the purchase queue whitelist
+     * @dev Only callable by the owner
+     * @param buyer The buyer to blacklist
+     */
+    function blacklistBuyer(address buyer) external override onlyOwner {
```

Comment:
Remove from whitelist does not necessarily mean blacklisting. This is slightly confusing since the buyer can still place purchase requests that meet the minimum requirement.